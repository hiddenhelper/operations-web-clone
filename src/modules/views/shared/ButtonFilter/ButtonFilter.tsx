import React, { memo, useCallback, useRef, useState } from 'react';

import Divider from '@material-ui/core/Divider';
import Button from '@material-ui/core/Button';

import ControlledButton from '../FormHandler/ControlledButton';

import { ExpandIcon } from '../../../../constants';
import { getConditionalDefaultValue, getDefaultValue } from '../../../../utils/generalUtils';
import { tableGlobalStyles } from '../../../../assets/styles';

export interface IButtonFilterProps {
  text: string;
  disabled?: boolean;
  showDivider?: boolean;
  transparentButton?: boolean;
  showIconMargin?: boolean;
  styleClass?: string;
  containerStyleClass?: string;
  buttonProps?: Object;
  render: (data?: any, other?: any) => React.ReactNode;
}

const ButtonFilter = ({
  buttonProps,
  text,
  showDivider,
  transparentButton = false,
  styleClass,
  containerStyleClass,
  disabled = false,
  render,
}: IButtonFilterProps) => {
  const tableGlobalClasses = tableGlobalStyles();

  const [isOpen, setIsOpen] = useState(false);
  const anchorRef = useRef();

  const handleToggle = useCallback(() => {
    setIsOpen(open => !open);
  }, []);
  const handleClose = useCallback(() => {
    setIsOpen(false);
  }, []);

  return (
    <div
      className={`${tableGlobalClasses.floatingFilterWrapper} ${tableGlobalClasses.filterWrapper} ${getDefaultValue(
        containerStyleClass,
        ''
      )} floatingFilterWrapper`}
    >
      {showDivider && <Divider orientation="vertical" flexItem={true} className={tableGlobalClasses.dividerNoSpacing} />}
      <ControlledButton styleClass={tableGlobalClasses.floatingFilterButton}>
        <Button
          className={getDefaultValue(styleClass, tableGlobalClasses.buttonFilter)}
          variant={getConditionalDefaultValue(transparentButton, 'text', 'contained')}
          onClick={handleToggle}
          disableRipple={true}
          aria-controls={isOpen && 'menu-list-grow'}
          aria-haspopup="true"
          disabled={disabled}
          endIcon={<ExpandIcon className={tableGlobalClasses.dropdownIcon} />}
          data-testid="button-filter-open"
          {...buttonProps}
        >
          {getDefaultValue(text, '')}
        </Button>
      </ControlledButton>
      {render({ isOpen, anchorRef, handleClose })}
    </div>
  );
};

export default memo(ButtonFilter);
