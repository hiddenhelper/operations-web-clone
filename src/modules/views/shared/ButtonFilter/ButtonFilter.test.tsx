import React from 'react';
import { render, act, fireEvent } from '@testing-library/react';

import ButtonFilter, { IButtonFilterProps } from './ButtonFilter';

describe('ButtonFilter Component', () => {
  let props: IButtonFilterProps;

  beforeEach(() => {
    props = {
      text: '',
      disabled: false,
      showDivider: true,
      styleClass: '',
      buttonProps: {},
      render: currentProps => {
        return (
          <div>
            <button data-testid="button-filter-close" onClick={() => currentProps.handleClose()}>
              Close
            </button>
            Children Node
          </div>
        );
      },
    };
  });

  it('should render', () => {
    const wrapper = render(<ButtonFilter {...props} />);
    expect(wrapper.container).toMatchSnapshot();
  });

  it('should toggle', () => {
    const wrapper = render(<ButtonFilter {...props} />);
    act(() => {
      fireEvent.click(wrapper.getByTestId('button-filter-open'));
    });
    expect(wrapper.container).toMatchSnapshot();
  });

  it('should close', () => {
    const wrapper = render(<ButtonFilter {...props} />);
    act(() => {
      fireEvent.click(wrapper.getByTestId('button-filter-close'));
    });
    expect(wrapper.container).toMatchSnapshot();
  });
});
