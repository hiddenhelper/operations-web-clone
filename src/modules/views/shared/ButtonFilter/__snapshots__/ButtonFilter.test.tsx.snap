// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonFilter Component should close 1`] = `
<div>
  <div
    class="makeStyles-floatingFilterWrapper-99 makeStyles-filterWrapper-98  floatingFilterWrapper"
  >
    <hr
      class="MuiDivider-root makeStyles-dividerNoSpacing-104 MuiDivider-flexItem MuiDivider-vertical"
    />
    <span
      class="makeStyles-floatingFilterButton-100"
      data-testid="controlled-button"
      tabindex="-1"
    >
      <button
        aria-controls="false"
        aria-haspopup="true"
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-buttonFilter-108"
        data-testid="button-filter-open"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          
          <span
            class="MuiButton-endIcon MuiButton-iconSizeMedium"
          >
            <svg
              class="makeStyles-dropdownIcon-102"
            >
              expand.svg
            </svg>
          </span>
        </span>
      </button>
    </span>
    <div>
      <button
        data-testid="button-filter-close"
      >
        Close
      </button>
      Children Node
    </div>
  </div>
</div>
`;

exports[`ButtonFilter Component should render 1`] = `
<div>
  <div
    class="makeStyles-floatingFilterWrapper-21 makeStyles-filterWrapper-20  floatingFilterWrapper"
  >
    <hr
      class="MuiDivider-root makeStyles-dividerNoSpacing-26 MuiDivider-flexItem MuiDivider-vertical"
    />
    <span
      class="makeStyles-floatingFilterButton-22"
      data-testid="controlled-button"
      tabindex="-1"
    >
      <button
        aria-controls="false"
        aria-haspopup="true"
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-buttonFilter-30"
        data-testid="button-filter-open"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          
          <span
            class="MuiButton-endIcon MuiButton-iconSizeMedium"
          >
            <svg
              class="makeStyles-dropdownIcon-24"
            >
              expand.svg
            </svg>
          </span>
        </span>
      </button>
    </span>
    <div>
      <button
        data-testid="button-filter-close"
      >
        Close
      </button>
      Children Node
    </div>
  </div>
</div>
`;

exports[`ButtonFilter Component should toggle 1`] = `
<div>
  <div
    class="makeStyles-floatingFilterWrapper-60 makeStyles-filterWrapper-59  floatingFilterWrapper"
  >
    <hr
      class="MuiDivider-root makeStyles-dividerNoSpacing-65 MuiDivider-flexItem MuiDivider-vertical"
    />
    <span
      class="makeStyles-floatingFilterButton-61"
      data-testid="controlled-button"
      tabindex="-1"
    >
      <button
        aria-controls="menu-list-grow"
        aria-haspopup="true"
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-buttonFilter-69"
        data-testid="button-filter-open"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          
          <span
            class="MuiButton-endIcon MuiButton-iconSizeMedium"
          >
            <svg
              class="makeStyles-dropdownIcon-63"
            >
              expand.svg
            </svg>
          </span>
        </span>
      </button>
    </span>
    <div>
      <button
        data-testid="button-filter-close"
      >
        Close
      </button>
      Children Node
    </div>
  </div>
</div>
`;
